python/rmm/_lib/device_buffer.pyx

```diff
@@ -109,22 +114,37 @@ cdef class DeviceBuffer:
     def frombytes(const unsigned char[::1] b, uintptr_t stream=0):
         return DeviceBuffer.c_frombytes(b, stream)
 
+    cpdef copy_to_host(self, unsigned char[::1] hb=None, uintptr_t stream=0):
+        cdef const device_buffer* dbp = self.c_obj.get()
+        cdef size_t s = dbp.size()
+
+        if hb is None:
+            hb = np.empty((s,), dtype="u1")
```

Comment:
I thought we wanted to return a bytes object instead of a numpy array here?

Reply:
As we are trying to match the semantics of Numba's `copy_to_host` and they use a NumPy array by default, that's what we do here. However it is possible to allocate a `bytes` object and fill it using this function. We do this in `tobytes` now for context.

Reply:
We are trying to move away from array semantics in rmm so I would be in favor of just having `tobytes` and making this an alias then. We can construct a numpy array zero copy from a bytes object.

Reply:
The trick with that is we will have different behavior from Numba's `copy_to_host` function. This will make it harder for people to rely on duck-typing to simply swap out the Numba implementation for the RMM one. Should add duck-typing is something we now rely on in dask-cuda and part of my motivation for writing this 😉

That all being said, this shouldn't stop you from coercing to `bytes` if you prefer. In fact am discussing with other projects if we can add `__bytes__` in more places ( https://github.com/numba/numba/issues/5154 ) ( https://github.com/cupy/cupy/issues/3007 ). Should add CuPy also has a `tobytes` function. If Numba added `to bytes` as well, we could also rely on duck-typing for this feature too.

Thoughts? 🙂

Reply:
I strongly prefer protocols over duck typing as APIs tend to change much more rapidly and freely than protocols (especially without ecosystem conversation). In general I would like to stop mirroring numba's API in general and rely on interop more and make rmm less of an array library and more of a buffer / memory allocation library.

Reply:
Generally I agree and am ok replacing these long term. In the short term am just interested in wiring this into cuDF and dask-cuda quickly to improve performance. This API just seemed like a short path to doing that. Happy to rip it out and replace it with something later. 😄

As to the particulars of using a NumPy array vs. using `bytes`, there seems to be a performance difference. After looking into this a bit more deeply it seems related to NumPy using huge pages. We can get comparable performance using `bytes` if we enable huge pages (a couple line change). Though it does raise a question about whether we should enable this for `bytes` objects and when we should disable using them on a `bytes` object (given this is general Python memory we don't own). For reference here's a short benchmark:

```python
In [1]: import numpy as np                                                      

In [2]: import rmm                                                              

In [3]: rmm.reinitialize(pool_allocator=True, 
   ...:                  initial_pool_size=int(2 * 2**30))                      
Out[3]: 0

In [4]: b = rmm.DeviceBuffer.to_device(int(2**30) * b"a")                       

In [5]: %timeit b.tobytes()                                                     
635 ms ± 3.09 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [6]: %timeit b.copy_to_host()                                                
366 ms ± 354 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```